source/mainnet/smart-contracts/tutorials/fungible-tokens/smart-contract.rst

```diff
@@ -0,0 +1,517 @@
+
+But first, you need another parameter to get the information about the tokens that will be burned. Create a struct, called ``BurnParams``; you will need ``token_id`` and the amount to be burned.
+
+.. code-block:: rust
+
+    #[derive(Serial, Deserial, SchemaType)]
+    struct BurnParams {
+
+        token_id: ContractTokenId,
+        amount: ContractTokenAmount,
+    }
+
+When you get the parameters, ensure the token exists with the ``ensure!`` and ``contains_token()`` function. Note that, when you call the ``burn()`` function, you need to emit the BurnEvent. For more detail check the CIS-2_ standard documentation.
+
+.. code-block:: rust
+
+    #[receive(
+        contract = "fungible-cis2",
+        name = "burn",
+        parameter = "BurnParams",
+        error = "ContractError",
+        enable_logger,
+        mutable
+    )]
+    fn contract_burn<S: HasStateApi>(
+        ctx: &impl HasReceiveContext,
+        host: &mut impl HasHost<State<S>, StateApiType = S>,
+        logger: &mut impl HasLogger,
+    ) -> ContractResult<()> {
+        // Get the contract owner
+        // let owner = ctx.owner();
+        // Get the sender of the transaction
+        let sender = ctx.sender();
+
+        // Parse the parameter.
+        let params: BurnParams = ctx.parameter_cursor().get()?;
+        let token_id = params.token_id;
+
+        let amount = params.amount;
+        ensure!(
+                self.contains_token(&token_id),
+                ContractError::InvalidTokenId
+            );
+        let (state, builder) = host.state_and_builder();
+
+        // can use the value to store it in the state.
+        let remaining_amount: ContractTokenAmount = state.burn(&token_id, amount, &sender)?;
+
+        // log burn event
+        logger.log(&Cis2Event::Burn(BurnEvent {
+            token_id,
+            amount,
+            owner: sender,
+        }))?;
+        Ok(())
+    }
+
+Now add the ``burn()`` implementation to the state functions as shown below. You will simply remove the amount for that particular token from the state but you need to take into account that the address that called this has a balance. As a control statement, in case someone wants to burn 0 tokens, check it and let them know burning nothing is successful by returning ``Ok(amount)`` where the amount equals zero.
+
+Then get the balance of the owner using its address. Since you will change the balance you need to have a mutable state. With the match statement, you will iterate the addresses from the ``address_state``, and when you find the matching one you will get the mutable balance. If it has enough balance to burn, you will subtract that amount from the mutable balance. Then state burn() function will return ``ContractResult()`` with the burned amount if it is successful; otherwise, it will prompt a CustomContractError::NoBalanceToBurn.
```

Comment:
```suggestion
Then get the balance of the owner using its address. Since you will change the balance you need to have a mutable state. With the match statement, you will iterate the addresses from the ``address_state``, and when you find the matching one you will get the mutable balance. If it has enough balance to burn, you will subtract that amount from the mutable balance. Then state burn() function will return ``ContractResult()`` with the burned amount if it is successful; otherwise, it will prompt a ``CustomContractError::NoBalanceToBurn``.
```